File: src/algorithms/binary-search/binary-search.c
#include "binary-search.h"

#include <stdio.h>

int binary_search(int* arr, size_t len, int v) {
    int lo = 0;
    int hi = len;

    while (lo != hi) {
        int m = lo + (hi - lo) / 2;
        if (arr[m] == v) {
            return 1;
        }

        if (v > arr[m]) {
            lo = m + 1;
        }

        if (v < arr[m]) {
            hi = m;
        }
    }
    return -1;
}

File: src/algorithms/binary_search.go
package algorithms

func BinarySearch(list []int, f int) bool {
	lo := 0
	hi := len(list)

	for lo < hi {
		n := lo + (hi-lo)/2
		if list[n] == f {
			return true
		} else if f > list[n] {
			lo = n + 1
		} else if f < list[n] {
			hi = n
		}

	}

	return false
}
